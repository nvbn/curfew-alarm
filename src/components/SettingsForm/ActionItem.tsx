import React from "react";
import { Text, TouchableOpacity } from "react-native";

import { isReady } from "../../utils/future";
import styles from "./styles";
import { ActionItemProps } from "./types";

/**
 * Settings action that only visible when an action hasn't been performed yet.
 */
const ActionItem = ({
  id,
  title,
  value,
  onChange,
}: ActionItemProps): JSX.Element | null => {
  if (value || !isReady(value)) {
    return null;
  }

  return (
    <TouchableOpacity
      testID={`action-item-touchable-${id}`}
      style={styles.item}
      onPress={() => onChange(true)}
    >
      <Text style={styles.title}>{title}</Text>
    </TouchableOpacity>
  );
};

export default ActionItem;
