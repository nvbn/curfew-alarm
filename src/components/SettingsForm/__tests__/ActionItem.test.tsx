/* eslint-disable @typescript-eslint/no-empty-function */
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";

import ActionItem from "../ActionItem";

describe("<ActionItem />", () => {
  test("displays item when action hasn't been performed", () => {
    const rendered = render(
      <ActionItem
        id="test"
        title="Test action item"
        type="action"
        value={false}
        onChange={() => {}}
      />,
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  test("displays nothing when action has been performed", () => {
    const rendered = render(
      <ActionItem
        id="test"
        title="Test action item"
        type="action"
        value={true}
        onChange={() => {}}
      />,
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  test("trigger onChange on press", () => {
    const onChange = jest.fn();

    const { getByTestId } = render(
      <ActionItem
        id="test"
        title="Test action item"
        type="action"
        value={false}
        onChange={onChange}
      />,
    );

    const touchable = getByTestId("action-item-touchable-test");
    fireEvent.press(touchable);

    expect(onChange.mock.calls.length).toBe(1);
  });
});
