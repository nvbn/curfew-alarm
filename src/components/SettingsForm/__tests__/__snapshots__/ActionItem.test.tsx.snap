// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionItem /> displays item when action hasn't been performed 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderBottomColor": "#fafafa",
      "borderBottomWidth": 1,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
      "paddingHorizontal": 10,
      "paddingVertical": 20,
    }
  }
  testID="action-item-touchable-test"
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    Test action item
  </Text>
</View>
`;

exports[`<ActionItem /> displays item when action isn't ready 1`] = `null`;

exports[`<ActionItem /> displays nothing when action has been performed 1`] = `null`;
